File: projectforge-business/src/main/java/org/projectforge/business/common/BaseUserGroupRightsDO.java
package org.projectforge.business.common;

import javax.persistence.Column;
import javax.persistence.MappedSuperclass;

import org.projectforge.framework.persistence.entities.DefaultBaseDO;

/**
 * Created by blumenstein on 18.07.17.
 */
@MappedSuperclass
public class BaseUserGroupRightsDO extends DefaultBaseDO
{
  private String fullAccessGroupIds, fullAccessUserIds;

  private String readonlyAccessGroupIds, readonlyAccessUserIds;

  private String minimalAccessGroupIds, minimalAccessUserIds;

  /**
   * Members of these groups have full read/write access to all entries of this object.
   *
   * @return the fullAccessGroupIds
   */
  @Column(name = "full_access_group_ids", nullable = true)
  public String getFullAccessGroupIds()
  {
    return fullAccessGroupIds;
  }

  /**
   * These users have full read/write access to all entries of this object.
   *
   * @param fullAccessGroupIds the fullAccessGroupIds to set
   * @return this for chaining.
   */
  public BaseUserGroupRightsDO setFullAccessGroupIds(final String fullAccessGroupIds)
  {
    this.fullAccessGroupIds = fullAccessGroupIds;
    return this;
  }

  /**
   * @return the fullAccessUserIds
   */
  @Column(name = "full_access_user_ids", nullable = true)
  public String getFullAccessUserIds()
  {
    return fullAccessUserIds;
  }

  /**
   * @param fullAccessUserIds the fullAccessUserIds to set
   * @return this for chaining.
   */
  public BaseUserGroupRightsDO setFullAccessUserIds(final String fullAccessUserIds)
  {
    this.fullAccessUserIds = fullAccessUserIds;
    return this;
  }

  /**
   * Members of these groups have full read-only access to all entries of this object.
   *
   * @return the readonlyAccessGroupIds
   */
  @Column(name = "readonly_access_group_ids", nullable = true)
  public String getReadonlyAccessGroupIds()
  {
    return readonlyAccessGroupIds;
  }

  /**
   * @param readonlyAccessGroupIds the readonlyAccessGroupIds to set
   * @return this for chaining.
   */
  public BaseUserGroupRightsDO setReadonlyAccessGroupIds(final String readonlyAccessGroupIds)
  {
    this.readonlyAccessGroupIds = readonlyAccessGroupIds;
    return this;
  }

  /**
   * These users have full read-only access to all entries of this object.
   *
   * @return the readonlyAccessUserIds
   */
  @Column(name = "readonly_access_user_ids", nullable = true)
  public String getReadonlyAccessUserIds()
  {
    return readonlyAccessUserIds;
  }

  /**
   * @param readonlyAccessUserIds the readonlyAccessUserIds to set
   * @return this for chaining.
   */
  public BaseUserGroupRightsDO setReadonlyAccessUserIds(final String readonlyAccessUserIds)
  {
    this.readonlyAccessUserIds = readonlyAccessUserIds;
    return this;
  }

  /**
   * Members of these group have read-only access to all entries of this object, but they can only see the event start
   * and stop time
   *
   * @return the minimalAccessGroupIds
   */
  @Column(name = "minimal_access_group_ids", nullable = true)
  public String getMinimalAccessGroupIds()
  {
    return minimalAccessGroupIds;
  }

  /**
   * @param minimalAccessGroupIds the minimalAccessGroupIds to set
   * @return this for chaining.
   */
  public BaseUserGroupRightsDO setMinimalAccessGroupIds(final String minimalAccessGroupIds)
  {
    this.minimalAccessGroupIds = minimalAccessGroupIds;
    return this;
  }

  /**
   * Members of this group have only access to the start and stop time, nothing else.
   *
   * @return the minimalAccessUserIds
   */
  @Column(name = "minimal_access_user_ids", nullable = true)
  public String getMinimalAccessUserIds()
  {
    return minimalAccessUserIds;
  }

  /**
   * @param minimalAccessUserIds the minimalAccessUserIds to set
   * @return this for chaining.
   */
  public BaseUserGroupRightsDO setMinimalAccessUserIds(final String minimalAccessUserIds)
  {
    this.minimalAccessUserIds = minimalAccessUserIds;
    return this;
  }
}

File: projectforge-business/src/main/kotlin/org/projectforge/business/common/BaseUserGroupRightsDO.kt
/////////////////////////////////////////////////////////////////////////////
//
// Project ProjectForge Community Edition
//         www.projectforge.org
//
// Copyright (C) 2001-2019 Micromata GmbH, Germany (www.micromata.com)
//
// ProjectForge is dual-licensed.
//
// This community edition is free software; you can redistribute it and/or
// modify it under the terms of the GNU General Public License as published
// by the Free Software Foundation; version 3 of the License.
//
// This community edition is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General
// Public License for more details.
//
// You should have received a copy of the GNU General Public License along
// with this program; if not, see http://www.gnu.org/licenses/.
//
/////////////////////////////////////////////////////////////////////////////

package org.projectforge.business.common

import org.projectforge.framework.persistence.entities.DefaultBaseDO
import org.projectforge.framework.persistence.user.entities.PFUserDO
import javax.persistence.Column
import javax.persistence.MappedSuperclass
import javax.persistence.Transient

/**
 * Base class for objects supporting user and group specific rights. You may define single group and user ids for the
 * different access types, such as owner, full access, readonly access and minimal access.
 */
@MappedSuperclass
abstract class BaseUserGroupRightsDO : DefaultBaseDO() {

    /**
     * The owner of this object.
     */
    @get:Transient
    abstract val owner: PFUserDO?


    /**
     * The id of the owner if exist, otherwise null.
     */
    val ownerId
        @Transient
        get() = owner?.id

    /**
     * Members of these groups have full read/write access to all entries of this object.
     */
    @get:Column(name = "full_access_group_ids", nullable = true)
    open var fullAccessGroupIds: String? = null

    @get:Column(name = "full_access_user_ids", nullable = true)
    open var fullAccessUserIds: String? = null

    /**
     * Members of these groups have full read-only access to all entries of this object.
     */
    @get:Column(name = "readonly_access_group_ids", nullable = true)
    open var readonlyAccessGroupIds: String? = null

    /**
     * These users have full read-only access to all entries of this object.
     */
    @get:Column(name = "readonly_access_user_ids", nullable = true)
    open var readonlyAccessUserIds: String? = null

    /**
     * Members of these group have read-only access to all entries of this object, but they can only see a minimal
     * set of the data of this object. This is used e. g. for calendar entries, where the users may only see the
     * start and end time of an event, but no information of details such as location, notes etc.
     */
    @get:Column(name = "minimal_access_group_ids", nullable = true)
    open var minimalAccessGroupIds: String? = null

    /**
     * Members of this group have only access to the start and stop time, nothing else.
     */
    @get:Column(name = "minimal_access_user_ids", nullable = true)
    open var minimalAccessUserIds: String? = null
}
